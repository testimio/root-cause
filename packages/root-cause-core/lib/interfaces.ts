import type { Page as PuppeteerPage } from 'puppeteer';
import type { Page as PlaywrightPage } from 'playwright';
import type { TestContextInterface } from './TestContext';
import type { TestEndStatus } from './attachInterfaces';
import type { InstrumentedFunctionResult } from '@testim/root-cause-types';

export type RootCausePage = PuppeteerPage | PlaywrightPage;

export interface AttachOptions {
  noResultsServer?: boolean;
  externalTestName?: string;
  resultsDirectory?: string;
}

export interface CommonEachHookArgs {
  testContext: TestContextInterface;
  fnName: string;
  proxyContext: any;
  rootPage: RootCausePage;
  args: any[];
  methodCallData: ProxiedMethodCallData[];
}

export type BeforeHookArgs = CommonEachHookArgs;

export type BeforeHook = (args: BeforeHookArgs) => Promise<void>;

export interface AfterHookArgs extends CommonEachHookArgs {
  instrumentedFunctionResult: InstrumentedFunctionResult<any, any>;
}

export type AfterHook = (args: AfterHookArgs) => Promise<void>;

export interface BeforeAllHookArgs {
  testContext: TestContextInterface;
  proxyContext: any;
  rootPage: RootCausePage;
}

export type BeforeAllHook = (args: BeforeAllHookArgs) => Promise<void>;

export interface AfterAllHookArgs {
  testContext: TestContextInterface;
  endStatus: TestEndStatus<unknown, unknown>;
}

export type AfterAllHook = (args: AfterAllHookArgs) => Promise<void>;

export interface IAutomationFrameworkInstrumentor {
  // register additional hooks on a context, this is implemented for
  // puppeteer/selenium
  // registerHooks(page: T, hooks: AutomationFrameworkHooks<T>): T;
  registerBeforeAllHook(hook: BeforeAllHook): void;
  registerAfterAllHook(hook: AfterAllHook): void;
  registerBeforeHook(hook: BeforeHook): void;
  registerAfterHook(hook: AfterHook): void;
  wrapWithProxy<T extends object>(proxiedObject: T): T;
  end(endStatus: TestEndStatus<unknown, unknown>): Promise<void>;
  pause(): void;
  resume(): void;
}

export interface ProxiedMethodCallData {
  creationFunction: string;
  selector?: string;
  text?: string;
  index?: number;
}
