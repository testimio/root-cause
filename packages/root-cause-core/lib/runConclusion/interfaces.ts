import { TestResultFile } from '@testim/root-cause-types';

export interface RootCauseRunResultEntry {
  singleResultDir: string;
  id: string;
  testData: TestResultFile;
}
