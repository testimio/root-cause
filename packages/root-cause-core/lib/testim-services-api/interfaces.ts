// generated with json2ts

interface TestimExecutionExecutionRunConfig {
  parallel: number;
  browser: string;
  runnerVersion: string;
  testimBranch: string;
  canaryMode: boolean;
  source: string;
  testPlans: any[];
  testLabels: any[];
  testNames: any[];
  testIds: any[];
  testConfigs: any[];
  testConfigIds: any[];
  browserTimeout: number;
  timeout: number;
  newBrowserWaitTimeout: number;
  tunnelPort: string;
  runnerMode: string;
  sessionType: string;
  companyId: string;
  testData?: any;
  testDataTotal?: any;
  testDataIndex?: any;
  testConfig: {}; // not too interesting
}

interface TestExecution {
  testId: string;
  status: string;
  name: string;
  resultId: string;
  isTestsContainer: boolean;
  testRunOverrideStatus: string;
  config: TestimExecutionExecutionRunConfig;
  success?: boolean;
  reason?: string;
  childTestResultIds?: string[];
  startTime: number;
  endTime: number;
  parentResultId?: string;
  show?: boolean;
}

export interface TestimBackendExecutionFormatSubsetForReporting {
  runId: string;
  execution: {
    [id: string]: {
      testId: string;
      name: string;
      success?: boolean;
      resultId: string;
      parentResultId?: string;
      isTestsContainer: boolean;
    };
  };
  startTime: number;
}
export interface TestimBackendExecutionInputFormat {
  runId: string;
  projectId: string;
  labels: string;
  startTime: number;
  endTime: number;
  execution: { [id: string]: TestExecution };
  status: string;
  config: TestimExecutionExecutionRunConfig;
  resultLabels: string[];
  remoteRunId: string;
  metadata: any;
  show: boolean;
}
