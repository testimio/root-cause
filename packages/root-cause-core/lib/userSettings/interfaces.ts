export interface PossibleUserSettings {
  features?: Partial<FeaturesSettings>;
}

export interface FeaturesSettings {
  /**
   * Take screenshot for every step
   *
   * Default to true, with jpeg 85
   */
  screenshots?: boolean | ScreenshotsSettings;

  /**
   * Record the browser's console
   *
   * @default true
   */
  console?: boolean;

  /**
   * Record network activity, make it available in the UI and as har file
   */
  networkLogs?: boolean;

  /**
   * When using jest integration, record jest expect calls as assertion steps
   */
  jestAssertions?: boolean;

  /**
   * Record the DOM content of the page for every step
   */
  html?: boolean;
}

export interface ResolvedSettings {
  features: {
    screenshots: false | Required<ScreenshotsSettings>;
    console: boolean;
    networkLogs: boolean;
    jestAssertions: boolean;
    html: boolean;
  };
}

export interface ScreenshotsSettings {
  format: 'png' | 'jpeg';

  /**
   * Quality of jpeg image, from 1 to 100
   *
   * @default 85
   */
  quality?: number;

  /**
   * When true, takes a screenshot of the full scrollable page, instead of the currently visible
   *
   * @default false
   */
  fullPage?: boolean;
}

type ContentType = string;

/**
 * placeholder, not in use yet
 */
export interface NetworkLogsSettings {
  /**
   * Save the response bodies into the har file, or only for specific content types
   *
   * @default false
   */
  includeResponseBody?: boolean | ContentType[];
}
