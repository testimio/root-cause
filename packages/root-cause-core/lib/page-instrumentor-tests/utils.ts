import type { RootCausePage } from '../interfaces';
import { PuppeteerPageHooker } from '../PuppeteerPageHooker';
import { TestContextMemory } from '../TestContextMemory';
import { createDateMocker } from '../utils';

export async function runPageInstrumentorTest<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage,
  testCode: (wrappedPage: TPage) => Promise<void>
): Promise<SetupPageInstrumentorForTestingReturn<TPage>['hooks']> {
  const { pageInstrumentor, wrappedPage, hooks } = await setupPageInstrumentorForTesting(
    testName,
    fullTestFileName,
    page
  );

  pageInstrumentor.start();

  await testCode(wrappedPage);

  await pageInstrumentor.end({
    success: true,
  });

  return hooks;
}

interface SetupPageInstrumentorForTestingReturn<TPage extends RootCausePage> {
  wrappedPage: TPage;
  pageInstrumentor: PuppeteerPageHooker;
  hooks: {
    beforeAllHook: jest.Mock<any, any>;
    afterAllHook: jest.Mock;
    beforeEachHook: jest.Mock;
    afterEachHook: jest.Mock;
  };
}

export async function setupPageInstrumentorForTesting<TPage extends RootCausePage>(
  testName: string,
  fullTestFileName: string,
  page: TPage
): Promise<SetupPageInstrumentorForTestingReturn<TPage>> {
  const testContext = new TestContextMemory(
    '/tmp',
    testName,
    `full ${testName}`,
    fullTestFileName,
    {
      screenshots: false,
      console: false,
      networkLogs: false,
      jestAssertions: false,
      html: false,
    },
    createDateMocker()
  );

  const pageInstrumentor = new PuppeteerPageHooker(testContext, page);

  const wrappedPage = pageInstrumentor.wrapWithProxy(page);

  const beforeAllHook = jest.fn();
  const afterAllHook = jest.fn();

  const beforeEachHook = jest.fn();
  const afterEachHook = jest.fn();

  pageInstrumentor.registerBeforeAllHook(beforeAllHook);
  pageInstrumentor.registerAfterAllHook(afterAllHook);
  pageInstrumentor.registerBeforeHook(beforeEachHook);
  pageInstrumentor.registerAfterHook(afterEachHook);

  return {
    wrappedPage,
    pageInstrumentor,
    hooks: {
      beforeAllHook,
      afterAllHook,
      beforeEachHook,
      afterEachHook,
    },
  };
}

const objPropToClean = ['testContext', 'proxyContext', 'rootPage', 'instrumentedFunctionResult'];
export function cleanJestMockFunctionCalls(calls: any[]): any[] {
  return calls.map((callArg) => {
    const cloned = { ...callArg };

    for (const prop of objPropToClean) {
      if (prop in cloned) {
        cloned[prop] = 'cleaned';
      }
    }

    return cloned;
  });
}
