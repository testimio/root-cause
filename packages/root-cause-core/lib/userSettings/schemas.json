{
  "$ref": "#/definitions/PossibleUserSettings",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "FeaturesSettings": {
      "additionalProperties": false,
      "properties": {
        "console": {
          "default": true,
          "description": "Record the browser's console",
          "type": "boolean"
        },
        "jestAssertions": {
          "description": "When using jest integration, record jest expect calls as assertion steps",
          "type": "boolean"
        },
        "networkLogs": {
          "description": "Record network activity, make it available in the UI and as har file",
          "type": "boolean"
        },
        "screenshots": {
          "anyOf": [
            {
              "type": "boolean"
            },
            {
              "$ref": "#/definitions/ScreenshotsSettings"
            }
          ],
          "description": "Take screenshot for every step\n\nDefault to true, with jpeg 85"
        },
        "html": {
          "default": false,
          "description": "Record the DOM content of the page for every step\n\nDefaults to false",
          "type": "boolean"
        }
      },
      "type": "object"
    },
    "PossibleUserSettings": {
      "additionalProperties": false,
      "properties": {
        "features": {
          "$ref": "#/definitions/FeaturesSettings"
        }
      },
      "type": "object"
    },
    "ScreenshotsSettings": {
      "additionalProperties": false,
      "properties": {
        "format": {
          "enum": [
            "png",
            "jpeg"
          ],
          "type": "string"
        },
        "fullPage": {
          "default": false,
          "description": "When true, takes a screenshot of the full scrollable page, instead of the currently visible",
          "type": "boolean"
        },
        "quality": {
          "default": 85,
          "description": "Quality of jpeg image, from 1 to 100",
          "type": "number"
        }
      },
      "required": [
        "format"
      ],
      "type": "object"
    }
  }
}
