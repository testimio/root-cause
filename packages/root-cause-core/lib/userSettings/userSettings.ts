import { cosmiconfig } from 'cosmiconfig';
import { CONFIG_MODULE_NAME } from '../consts';
import Ajv from 'ajv';
import betterAjvErrors from 'better-ajv-errors';

import {
  PossibleUserSettings,
  ResolvedSettings,
  FeaturesSettings,
  ScreenshotsSettings,
} from './interfaces';
import debug from 'debug';
// The schema is generated from ts interfaces by running yarn generate-settings-json-schema
import userSettingsSchema from './schemas.json';
import { malformedConfigFile } from '../userFacingStrings/userFacingStrings';

const logger = debug('root-cause:user-settings');

export async function loadSettings(startFrom?: string): Promise<ResolvedSettings> {
  const configFromFile = await readUserConfigFromFile(startFrom);

  return resolveSettings(configFromFile || {});
}

export async function readUserConfigFromFile(
  startFrom?: string
): Promise<PossibleUserSettings | null> {
  const ajvInstance = new Ajv({ allErrors: true, jsonPointers: true });
  try {
    const result = await cosmiconfig(CONFIG_MODULE_NAME, {
      transform(result) {
        return result;
      },
    }).search(startFrom);

    if (!result || result.isEmpty) {
      logger('config file was empty');
      logger('file path ', result?.filepath);
      return null;
    }

    logger('Found config file', result?.filepath);

    const valid = ajvInstance.validate(userSettingsSchema, result.config);

    if (!valid) {
      // We may consider here to fail the invocation and not fallback to default config
      // eslint-disable-next-line no-console
      console.warn(malformedConfigFile());
      const betterErrors = betterAjvErrors(userSettingsSchema, result.config, ajvInstance.errors);
      // eslint-disable-next-line no-console
      console.log(betterErrors);
      // eslint-disable-next-line no-console
      console.log('-'.repeat(5));
    }

    return result.config;
  } catch (configSearchError) {
    // We may consider here to fail the invocation and not fallback to default config
    logger('cosmiconfig error', configSearchError);
    return null;
  }
}

export function resolveSettings(input: PossibleUserSettings): ResolvedSettings {
  const screenshots = resolveScreenshots(input.features?.screenshots);
  // networkLogs on by default
  const networkLogs = input.features?.networkLogs ?? true;
  // on by default
  const console = input.features?.console ? input.features?.console : true;
  // off by default
  const jestAssertions = input.features?.jestAssertions ? input.features?.jestAssertions : false;

  const features: ResolvedSettings['features'] = {
    screenshots,
    console,
    networkLogs,
    jestAssertions,
  };

  return {
    features,
  };
}

function resolveScreenshots(
  input: FeaturesSettings['screenshots']
): false | Required<ScreenshotsSettings> {
  if (input === true || input === undefined) {
    return {
      format: 'jpeg',
      quality: 85,
      fullPage: false,
    };
  }

  if (input === false) {
    return false;
  }

  return {
    quality: 85,
    fullPage: false,
    ...input,
  };
}

// function resolveNetworkLogs(input: FeaturesSettings['networkLogs']): false | Required<NetworkLogsSettings> {
//     if (input === undefined) {
//         return {
//             includeResponseBody: false,
//         };
//     }

//     if (input === false) {
//         return false;
//     }

//     if (input === true) {
//         return {
//             includeResponseBody: true,
//         };
//     }

//     return {
//         includeResponseBody: input.includeResponseBody ?? false,
//     };
// }
