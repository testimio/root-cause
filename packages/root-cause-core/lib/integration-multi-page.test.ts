import {
  getCleanAllPathsPrettyFormatPlugin,
  getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin,
} from '@testim/internal-self-tests-helpers';
// eslint-disable-next-line import/no-extraneous-dependencies
import { ensurePrerequisite, getCurrentTest } from '@testim/root-cause-jest';
import type { TestResultFile } from '@testim/root-cause-types';
import assert from 'assert';
import fs from 'fs-extra';
import path from 'path';
import puppeteer from 'puppeteer';
import { attach } from './index';
import { PossibleUserSettings } from './userSettings/interfaces';
import { resolveSettings } from './userSettings/userSettings';
import { jsonReduceNoiseReviver, testUniqueIdentifierFromStartParams } from './utils';

describe('integration multi page test', () => {
  jest.setTimeout(30_000);
  expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
  expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());

  let browser: puppeteer.Browser;
  let browserContext: puppeteer.BrowserContext;
  let page: puppeteer.Page;

  beforeAll(async () => {
    ensurePrerequisite();

    // https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
    // Not optimal, but didn't work on circle on circleci/node:12.17-stretch-browsers without it
    // Need to revisit
    browser = await puppeteer.launch({
      args: ['--no-sandbox', '--disable-setuid-sandbox'],
      devtools: false,
    });
  });

  afterAll(async () => {
    await browser.close();
  });

  beforeEach(async () => {
    browserContext = await browser.createIncognitoBrowserContext();
    page = await browserContext.newPage();
  });

  afterEach(async () => {
    await page.close();
    await browserContext.close();
  });

  it('integration multi page test', async () => {
    const currentTestInfo = getCurrentTest();

    const startTestParams = {
      runId: 'mock_invocation_id',
      projectRoot: path.resolve(__dirname, 'testsResults'),
      fullName: currentTestInfo.fullName,
      description: currentTestInfo.description,
      fullSuitePath: __filename,
    };

    const mockedDateConstructor: typeof Date = Object.create(Date);

    let nowCallsCounter = 1;
    mockedDateConstructor.now = function mockedNow() {
      return nowCallsCounter++;
    };

    const overriddenFeatures: PossibleUserSettings = {
      features: {
        html: true,
      },
    };

    const { page: playedPage, endTest } = await attach(
      {
        page,
        startTestParams,
        activeFeatures: resolveSettings(overriddenFeatures).features,
      },
      mockedDateConstructor
    );
    await playedPage.setContent(
      '<h1>PAGE 1</h1><a href="about:blank" target="blank">CLICK HERE</a>'
    );
    const [newPageOpenedViaLink] = await Promise.all([
      new Promise<puppeteer.Page>((res) => {
        playedPage.once('popup', res);
      }),
      page.click('a'),
    ]);
    await newPageOpenedViaLink.setContent('<h1>PAGE 2</h1>');
    await newPageOpenedViaLink.click('body');
    await playedPage.click('body');

    await endTest({
      success: false,
      error: { message: 'mocked error test failed' },
    });

    const expectedResultsPath = path.resolve(
      __dirname,
      'testsResults',
      '.root-cause',
      'runs',
      'mock_invocation_id',
      testUniqueIdentifierFromStartParams(startTestParams)
    );
    const testResults: TestResultFile = JSON.parse(
      await fs.readFile(path.resolve(expectedResultsPath, 'results.json'), 'utf-8'),
      jsonReduceNoiseReviver
    );

    expect(testResults).toMatchSnapshot();

    for (const stepResult of testResults.steps) {
      // Do also images blob comparison? TBH
      if (stepResult.screenshot) {
        assert.equal(
          await fs.pathExists(path.resolve(expectedResultsPath, stepResult.screenshot)),
          true,
          `${stepResult.screenshot} is missing`
        );
      }
    }
  });
});
