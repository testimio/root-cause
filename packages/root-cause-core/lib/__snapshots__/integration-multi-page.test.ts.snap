// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration multi page test integration multi page test 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 11,
    "fileName": "noise_removed/packages/root-cause-core/lib/integration-multi-page.test.ts",
    "hasNetworkLogs": true,
    "systemInfo": Object {
      "automationFramework": "puppeteer",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "HeadlessChrome/86.0.4240.0",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 600,
        "width": 800,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "codeLocationDetails": Object {
        "callstack": Array [
          Object {
            "column": 29,
            "file": "lib/assortedHooks.ts",
            "function": "testEndHook",
            "line": 116,
          },
          Object {
            "column": 15,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.end",
            "line": 229,
          },
          Object {
            "column": 5,
            "file": "lib/attach.ts",
            "function": "endTest",
            "line": 115,
          },
          Object {
            "column": 5,
            "file": "lib/integration-multi-page.test.ts",
            "function": "Object.<anonymous>",
            "line": 97,
          },
        ],
        "codeLines": Array [
          "    await newPageOpenedViaLink.click('body');",
          "    await playedPage.click('body');",
          "",
          "    await endTest({",
          "      success: false,",
          "      error: { message: 'mocked error test failed' },",
          "    });",
        ],
        "column": 5,
        "fromRowNumber": 94,
        "row": 97,
        "sourceFileRelativePath": "lib/integration-multi-page.test.ts",
        "toRowNumber": 100,
      },
      "error": Object {
        "message": "mocked error test failed",
        "name": "",
        "stack": "",
      },
      "success": false,
    },
    "testFullName": "integration multi page test integration multi page test",
    "testName": "integration multi page test",
    "timestamp": 2,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 4,
      "fnName": "setContent",
      "index": 1,
      "mhtmlFile": "1.document.mhtml",
      "name": "setContent",
      "pageId": 0,
      "screenshot": "1.page-setContent.jpg",
      "startTimestamp": 3,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_setContent",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/integration-multi-page.test.ts",
            "function": "Object.<anonymous>",
            "line": 84,
          },
        ],
        "codeLines": Array [
          "      },",
          "      mockedDateConstructor",
          "    );",
          "    await playedPage.setContent(",
          "      '<h1>PAGE 1</h1><a href=\\"about:blank\\" target=\\"blank\\">CLICK HERE</a>'",
          "    );",
          "    const [newPageOpenedViaLink] = await Promise.all([",
        ],
        "column": 5,
        "fromRowNumber": 81,
        "row": 84,
        "sourceFileRelativePath": "lib/integration-multi-page.test.ts",
        "toRowNumber": 87,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 6,
      "fnName": "setContent",
      "index": 2,
      "mhtmlFile": "2.document.mhtml",
      "name": "setContent",
      "pageId": 1,
      "screenshot": "2.page-setContent.jpg",
      "startTimestamp": 5,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_setContent",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/integration-multi-page.test.ts",
            "function": "Object.<anonymous>",
            "line": 93,
          },
        ],
        "codeLines": Array [
          "      }),",
          "      page.click('a'),",
          "    ]);",
          "    await newPageOpenedViaLink.setContent('<h1>PAGE 2</h1>');",
          "    await newPageOpenedViaLink.click('body');",
          "    await playedPage.click('body');",
          "",
        ],
        "column": 5,
        "fromRowNumber": 90,
        "row": 93,
        "sourceFileRelativePath": "lib/integration-multi-page.test.ts",
        "toRowNumber": 96,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 8,
      "fnName": "click",
      "index": 3,
      "mhtmlFile": "3.document.mhtml",
      "name": "click \\"body\\"",
      "pageId": 1,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "body",
      "startTimestamp": 7,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/integration-multi-page.test.ts",
            "function": "Object.<anonymous>",
            "line": 94,
          },
        ],
        "codeLines": Array [
          "      page.click('a'),",
          "    ]);",
          "    await newPageOpenedViaLink.setContent('<h1>PAGE 2</h1>');",
          "    await newPageOpenedViaLink.click('body');",
          "    await playedPage.click('body');",
          "",
          "    await endTest({",
        ],
        "column": 5,
        "fromRowNumber": 91,
        "row": 94,
        "sourceFileRelativePath": "lib/integration-multi-page.test.ts",
        "toRowNumber": 97,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 10,
      "fnName": "click",
      "index": 4,
      "mhtmlFile": "4.document.mhtml",
      "name": "click \\"body\\"",
      "pageId": 0,
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "body",
      "startTimestamp": 9,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 196,
          },
          Object {
            "column": 5,
            "file": "lib/integration-multi-page.test.ts",
            "function": "Object.<anonymous>",
            "line": 95,
          },
        ],
        "codeLines": Array [
          "    ]);",
          "    await newPageOpenedViaLink.setContent('<h1>PAGE 2</h1>');",
          "    await newPageOpenedViaLink.click('body');",
          "    await playedPage.click('body');",
          "",
          "    await endTest({",
          "      success: false,",
        ],
        "column": 5,
        "fromRowNumber": 92,
        "row": 95,
        "sourceFileRelativePath": "lib/integration-multi-page.test.ts",
        "toRowNumber": 98,
      },
      "unhandledExceptions": Array [],
    },
  ],
}
`;
